Fix #1352: Allow X with ndims greater than 2

taharallouche opened the pull request:
## Description
Hello, this PR fixes #1352.

Just like #898, this issue shows that the call to check_array is unnecessarily too restrictive.

This is a low hanging fruit that would hopefully make the library usable for more applications :chart_with_upwards_trend:

I'll add a changelog entry if the fix is accepted :pray: 

-------

PS: the error message that `check_array` outputs is not user-friendly in absence of an `estimator` argument:
> ValueError: Found array with dim 4. None expected <= 2.

I'll try to open a PR to fix it upstream :sweat_smile: 


## Tests
<!--- Select all that apply by putting an x between the brackets: [x] -->
- [ ] no new tests required
- [x] new tests added
- [x] existing tests adjusted

## Documentation
<!--- Select all that apply. -->
- [x] no documentation changes needed
- [ ] user guide added or updated
- [ ] API docs added or updated
- [ ] example notebook added or updated

taharallouche commented on fairlearn/utils/_input_validation.py:
```diff
@@ -125,12 +125,7 @@ def _validate_and_reformat_input(
-        sensitive_features,
-        control_features,
-    )
+    return (X, result_y, sensitive_features, control_features)
```
`X` can no longer be converted to a `DataFrame`, because it's no longer guaranteed to be tabular.

I double checked that the returned `X` of this function is never used in the code-base so this change is safe.

TamaraAtanasoska replied:
I am wondering if it is safer for now to preserve the type? If a dataframe was entered then it is also returned again. It  will make it safer to merge and an easy fix if (more like when) we have to play with the internal pandas representations.

TamaraAtanasoska commented:
I just pushed to this file recently for the first when I was making fairlearn `sklearn 1.6` compatible and it is definitely in a need for refactoring 🙈. 

Generally LGTM, checking some things around the `Dataframe` type change and how it relates to something I am working on right now, and I'll do an official review in the next two days.

taharallouche commented:
Hi @TamaraAtanasoska, I put back the `DataFrame` type, added a test case to ensure `ExponentiatedGradient` can handle multi-dimensional data (which it doesn't on `main`), and added a changelog entry.

TamaraAtanasoska reviewed (approved):
Thanks @taharallouche LGTM! I can't merge it atm because of the persisting OpenML outage, we'd have to wait a bit.
